python/ray/llm/_internal/batch/processor/base.py

```diff
@@ -0,0 +1,194 @@
+    )
+
+    class Config:
+        validate_assignment = True
+        arbitrary_types_allowed = True
+
+
+@DeveloperAPI
+class Processor:
+    """A processor is composed of a preprocess stage, followed by one or more
+    processing stages, and finally a postprocess stage. We use processor as a
+    paradigm for processing data using LLMs.
+
+    Args:
+        config: The processor config.
+        preprocess: A lambda function that takes a row (dict) as input and returns a
+            preprocessed row (dict). The output row must contain the required fields
+            for the following processing stages.
+        postprocess: A lambda function that takes a row (dict) as input and returns a
+            postprocessed row (dict).
+        accelerator_type: The accelerator type. Default to None, meaning that only
+            the CPU will be used.
+        concurrency: The number of workers for data parallelism.
+    """
+
+    # The internal used data column name ("__data"). Your input
+    # dataset should not contain this column. If you want to use this column
+    # in your input dataset, you have to derive and customize Processor.
+    data_column: str = "__data"
+
+    def __init__(
+        self,
+        config: ProcessorConfig,
+        preprocess: Optional[UserDefinedFunction] = None,
+        postprocess: Optional[UserDefinedFunction] = None,
+        accelerator_type: Optional[str] = None,
+        concurrency: int = 1,
+    ):
+        self.config = config
+        self.preprocess = None
+        self.postprocess = None
+        self.accelerator_type = accelerator_type
+        self.concurrency = concurrency
+        self.stages: OrderedDict[str, StatefulStage] = OrderedDict()
+
+        if preprocess is not None:
+            self.preprocess = wrap_preprocess(
+                preprocess,
+                self.data_column,
+            )
+
+        if postprocess is not None:
+            self.postprocess = wrap_postprocess(
+                postprocess,
+                self.data_column,
+            )
+
+    def __call__(self, dataset: Dataset) -> Dataset:
+        """Execute the processor:
+        preprocess -> stages -> postprocess.
+        Note that the dataset won't be materialized during the execution.
+
+        Args:
+            dataset: The input dataset.
+
+        Returns:
+            The output dataset.
+        """
+        if self.preprocess is not None:
+            dataset = dataset.map(self.preprocess)
+
+        # Apply stages.
+        for stage in self.stages.values():
+            kwargs = stage.get_dataset_map_batches_kwargs(
+                batch_size=self.config.batch_size,
+                data_column=self.data_column,
+            )
+            dataset = dataset.map_batches(stage.fn, **kwargs)
+
+        if self.postprocess is not None:
+            dataset = dataset.map(self.postprocess)
+        return dataset
+
+    def append_stage(self, stage: StatefulStage) -> None:
```

Comment:
I'd like to make the Processor class immutable if possible. Can we append stage before creating the Processor object?

Reply:
Because `Processor` is a user-facing API. Exposing this API would make the users think they can/should append something into an existing processor if they have a different requirement. 
But in fact, the recommended  way is to define a different processor type and config. 
So not exposing this API can avoid this confusion and make it less error prone. 
If someone updates the interval `stages` variable, that's out of our control. We don't need to handle that case.